File: Array/1380. Lucky Numbers in a Matrix.cpp
class Solution {
public:
    vector<int> luckyNumbers (vector<vector<int>>& matrix) {
        // Time Complexity: O(mn)
        // Space Complexity: O(m + n)
        int m(matrix.size()), n(matrix[0].size());
		// record min of each row and max of each column
        vector<int> mins(m, 100001), maxs(n);
        vector<int> lucky;
        for (int i = 0; i < m; ++i)
            for (int j = 0; j < n; ++j)
                mins[i] = min(mins[i], matrix[i][j]);
        for (int j = 0; j < n; ++j)
            for (int i = 0; i < m; ++i)
                maxs[j] = max(maxs[j], matrix[i][j]);
        for (int i = 0; i < m; ++i)
            for (int j = 0; j < n; ++j)
                if (maxs[j] == matrix[i][j] && mins[i] == matrix[i][j])
                    lucky.push_back(matrix[i][j]);
        return lucky;
    }
};



//set Interaction
class Solution {
public:
    vector<int> luckyNumbers (vector<vector<int>>& matrix) {
        // Time Complexity: O(mn)
        // Space Complexity: O(min(m, n)) <- slightly improved, not max(m, n)
        int m(matrix.size()), n(matrix[0].size());
        unordered_set<int> extrema;
        vector<int> lucky;
        if (m < n) {
            for (int i = 0; i < m; ++i) {
                int r(100001);
                for (int j = 0; j < n; ++j)
                    r = min(r, matrix[i][j]);
                extrema.insert(r);
            }
            for (int j = 0; j < n; ++j) {
                int c(0);
                for (int i = 0; i < m; ++i)
                    c = max(c, matrix[i][j]);
                if (extrema.count(c)) lucky.push_back(c);
            }
        } else {
            for (int j = 0; j < n; ++j) {
                int c(0);
                for (int i = 0; i < m; ++i)
                    c = max(c, matrix[i][j]);
                extrema.insert(c);
            }
            for (int i = 0; i < m; ++i) {
                int r(100001);
                for (int j = 0; j < n; ++j)
                    r = min(r, matrix[i][j]);
                if (extrema.count(r)) lucky.push_back(r);
            }
        }
        
        return lucky;
    }
};

File: Array/1380. Lucky Numbers in a Matrix.py
class Solution:
    def luckyNumbers (self, matrix: List[List[int]]) -> List[int]:
        if len(matrix) == 0 or len(matrix[0]) == 0:
            return 0
        l = collections.defaultdict(int)
        n, m = len(matrix), len(matrix[0])
        def findMax(j):
            if j in l:
                return l[j]
            maxn = matrix[0][j]
            for i in range(1,n):
                if matrix[i][j] > maxn:
                        maxn = matrix[i][j] 
            l[j] = maxn
            return maxn
        res = []
        for r in matrix:
            minn = min(r)
            ind = r.index(minn)
            if findMax(ind) == minn:
                res.append(minn)
        return res

“”“
We can create a set of the minimum row values using set comprehension.
To create a set of the maximum column values we can use the unpacking operator (*) on the matrix 
        and use zip to iterate over the unpacked rows in parallel.
We can then get a set of our lucky numbers by intersecting those two sets using the intersection operator (&). 
    Then all that is left to do is to convert the set back into a list.

”“”

class Solution:
    def luckyNumbers(self, matrix: List[List[int]]) -> List[int]:
        return list({min(row) for row in matrix} & {max(col) for col in zip(*matrix)})